test/core/event_engine/cf/cf_engine_test.cc

```diff
@@ -80,6 +82,211 @@ TEST(CFEventEngineTest, TestConnectionCancelled) {
 }
 
+namespace {
+std::vector<std::string> ResolvedAddressesToStrings(
+    const std::vector<EventEngine::ResolvedAddress> addresses) {
+  std::vector<std::string> ip_strings;
+  std::transform(addresses.cbegin(), addresses.cend(),
+                 std::back_inserter(ip_strings), [](auto const& address) {
+                   return ResolvedAddressToString(address).value_or("ERROR");
+                 });
+  return ip_strings;
+}
+}  // namespace
+
+TEST(CFEventEngineTest, CFEventEngineTest_TestCreateDNSResolver_Test) {
+  grpc_core::MemoryQuota memory_quota("cf_engine_test");
+  auto cf_engine = std::make_shared<CFEventEngine>();
+
+  EXPECT_TRUE(cf_engine->GetDNSResolver({}).status().ok());
+  EXPECT_TRUE(cf_engine->GetDNSResolver({.dns_server = ""}).status().ok());
+  EXPECT_EQ(
+      cf_engine->GetDNSResolver({.dns_server = "8.8.8.8"}).status().code(),
+      absl::StatusCode::kInvalidArgument);
+  EXPECT_EQ(
+      cf_engine->GetDNSResolver({.dns_server = "8.8.8.8:53"}).status().code(),
+      absl::StatusCode::kInvalidArgument);
+  EXPECT_EQ(
+      cf_engine->GetDNSResolver({.dns_server = "invalid"}).status().code(),
+      absl::StatusCode::kInvalidArgument);
+}
+
+TEST(CFEventEngineTest, TestResolveLocalhost) {
+  grpc_core::Notification resolve_signal;
+
+  auto cf_engine = std::make_shared<CFEventEngine>();
+  auto dns_resolver = cf_engine->GetDNSResolver({});
+
+  dns_resolver.value()->LookupHostname(
+      [&resolve_signal](auto result) {
+        EXPECT_TRUE(result.status().ok());
+        EXPECT_THAT(ResolvedAddressesToStrings(result.value()),
+                    testing::UnorderedElementsAre("127.0.0.1:80", "[::1]:80"));
+
+        resolve_signal.Notify();
+      },
+      "localhost", "80");
+
+  resolve_signal.WaitForNotification();
+}
+
+TEST(CFEventEngineTest, TestResolveRemote) {
+  grpc_core::Notification resolve_signal;
+
+  auto cf_engine = std::make_shared<CFEventEngine>();
+  auto dns_resolver = cf_engine->GetDNSResolver({});
+
+  dns_resolver.value()->LookupHostname(
+      [&resolve_signal](auto result) {
+        EXPECT_TRUE(result.status().ok());
+        EXPECT_THAT(ResolvedAddressesToStrings(result.value()),
+                    testing::UnorderedElementsAre("127.0.0.1:80", "[::1]:80"));
+
+        resolve_signal.Notify();
+      },
+      "localtest.me:80", "443");
+
+  resolve_signal.WaitForNotification();
+}
+
+TEST(CFEventEngineTest, TestResolveIPv4Remote) {
+  grpc_core::Notification resolve_signal;
+
+  auto cf_engine = std::make_shared<CFEventEngine>();
+  auto dns_resolver = cf_engine->GetDNSResolver({});
+
+  dns_resolver.value()->LookupHostname(
+      [&resolve_signal](auto result) {
+        EXPECT_TRUE(result.status().ok());
+        EXPECT_THAT(ResolvedAddressesToStrings(result.value()),
+                    testing::IsSubsetOf(
+                        {"1.2.3.4:80", "[64:ff9b::102:304]:80" /*NAT64*/}));
+
+        resolve_signal.Notify();
+      },
+      "1.2.3.4.nip.io:80", "");
```

Comment:
Same as above. Word of caution that any test that relies on an external service may A) become flaky even though the implementation is fine, and B) may be pretty unfriendly if we run `bazel test --runs_per_test=1000 ....`